src/integrations/kafka/consumer.cpp

```diff
@@ -450,4 +470,40 @@ void Consumer::ConsumerRebalanceCb::rebalance_cb(RdKafka::KafkaConsumer *consume
   }
 }
 void Consumer::ConsumerRebalanceCb::set_offset(int64_t offset) { offset_ = offset; }
+
+void Consumer::CheckAndDestroyLastAssignmentIfNeeded() const {
+  if (!last_assignment_.empty()) {
+    if (const auto err = consumer_->assign(last_assignment_); err != RdKafka::ERR_NO_ERROR) {
+      throw ConsumerStartFailedException(info_.consumer_name,
+                                         fmt::format("Couldn't restore commited offsets: '{}'", RdKafka::err2str(err)));
+    }
+    RdKafka::TopicPartition::destroy(last_assignment_);
+  }
+}
+
+bool Consumer::TryToConsumeBatch(const std::vector<Message> &batch) const {
+  try {
+    consumer_function_(batch);
+    std::vector<RdKafka::TopicPartition *> partitions;
+    utils::OnScopeExit clear_partitions([&]() { RdKafka::TopicPartition::destroy(partitions); });
+
+    if (const auto err = consumer_->assignment(partitions); err != RdKafka::ERR_NO_ERROR) {
+      throw ConsumerCheckFailedException(
+          info_.consumer_name, fmt::format("Couldn't get assignment to commit offsets: {}", RdKafka::err2str(err)));
+    }
+    if (const auto err = consumer_->position(partitions); err != RdKafka::ERR_NO_ERROR) {
+      throw ConsumerCheckFailedException(info_.consumer_name,
+                                         fmt::format("Couldn't get offsets from librdkafka {}", RdKafka::err2str(err)));
+    }
+    if (const auto err = consumer_->commitSync(partitions); err != RdKafka::ERR_NO_ERROR) {
+      spdlog::warn("Committing offset of consumer {} failed: {}", info_.consumer_name, RdKafka::err2str(err));
+      return false;
+    }
+  } catch (const std::exception &e) {
```

Comment:
And of course I wouldn't catch the exceptions here, but only in `StartConsuming`. In `StartConsumingWithLimit` you can leave the exceptions to propagate into the query engine to make it visible to the users.